src/providers/pkcs11/capability_discovery.rs

```diff
@@ -0,0 +1,126 @@
+            CheckType::Generate => return Provider::generate_check(attributes),
+            CheckType::Import => return Provider::import_check(attributes),
+            CheckType::Use => {
+                if attributes.policy.permitted_algorithms == Algorithm::None {
+                    return Err(PsaErrorNotSupported);
+                } else {
+                    return Provider::use_check(&self, attributes);
+                }
+            }
+            CheckType::Derive => return Provider::derive_check(attributes),
+        };
+        return Ok(can_do_crypto::Result {});
+    }
+
+    fn use_check(&self, attributes: Attributes) -> Result<can_do_crypto::Result> {
+        if !(attributes.policy.usage_flags.decrypt()
+            || attributes.policy.usage_flags.encrypt()
+            || attributes.policy.usage_flags.sign_hash()
+            || attributes.policy.usage_flags.sign_message()
+            || attributes.policy.usage_flags.verify_hash()
+            || attributes.policy.usage_flags.verify_message())
+        {
+            return Err(PsaErrorNotSupported);
+        }
+        attributes.compatible_with_alg(attributes.policy.permitted_algorithms)?;
+        let supported_mechanisms: Vec<MechanismType> = self
+            .backend
+            .get_mechanism_list(self.slot_number)
+            .map_err(to_response_status)?;
+        let mechanism = Mechanism::try_from(attributes.policy.permitted_algorithms)
+            .map_err(to_response_status)?;
+        let mechanism_info: MechanismInfo = self
+            .backend
+            .get_mechanism_info(self.slot_number, mechanism.mechanism_type())
+            .map_err(to_response_status)?;
+        if !(supported_mechanisms.contains(&mechanism.mechanism_type())) {
+            return Err(PsaErrorNotSupported);
+        }
+        if std::any::type_name::<Ulong>() == std::any::type_name::<u64>() {
+            if !((attributes.bits as u64) >= (*mechanism_info.min_key_size()).into()
+                && (attributes.bits as u64) <= (*mechanism_info.max_key_size()).into())
+            {
+                return Err(PsaErrorNotSupported);
+            }
+        } else {
+            if !((*mechanism_info.min_key_size() as u64) <= (attributes.bits as u64)
+                && (attributes.bits as u64) <= (*mechanism_info.max_key_size() as u64))
+            {
+                return Err(PsaErrorNotSupported);
+            }
+        }
+        return Ok(can_do_crypto::Result {});
+    }
+
+    fn derive_check(attributes: Attributes) -> Result<can_do_crypto::Result> {
+        return Err(PsaErrorNotSupported);
+    }
+
+    fn generate_check(attributes: Attributes) -> Result<can_do_crypto::Result> {
+        if !(attributes.key_type == Type::RsaKeyPair
+            || attributes.key_type
+                == (Type::EccKeyPair {
+                    curve_family: EccFamily::SecpR1,
+                }))
+        {
+            return Err(PsaErrorNotSupported);
+        }
```

Comment:
If `attributes.policy.permitted_algorithms` is not `None`, I think we should call `use_check` from here too. Same for `import`. I think some tests with non-compatible key type/algorithm are still passing because of that.

Reply:
Calling use_check causes an issue as it checks if at least on of the the relevant usage flags are set which isn't applicable to import and generate so would it work to just run `attributes.compatible_with_alg(attributes.policy.permitted_algorithms)?;` to check if the type and algorithm match?

Reply:
Reading the [contract](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/can_do_crypto.html#description) we have written now, I think it should be applicable?

For the Generate, Import and Derive check there is:

> If the `key_algorithm` is not `None`, also perform the `Use` check.

Reply:
So are we saying that Generate and Import should have also have at least one of the relevant usage flags set? (And Derive but that is never supported by pkcs11)

Reply:
I think so, if an Algorithm is set, it means that it should be used for at least something